### PR TITLE
#161255206 Users should be able to follow each other

### PR DESCRIPTION
**Pull request by mmosoroohh**
#### What does this PR do?
 - Users should be able to follow and or unfollow other users.
#### Description of Task to be completed?
- Create a follow button component
- Link follow button on user card on an article
#### How should this be manually tested?
Clone this repository.
- Branch out to `ft-users-able-follow-each-other-161255206`
- Install node and yarn
- Run `yarn install`
- Run `yarn start`

#### What are the relevant pivotal tracker stories?
[#161255206](https://www.pivotaltracker.com/story/show/161255206)

#### Screenshots (if appropriate)
- Follow a user
screenshot 2018-12-19 at 23 34 11

- Unfollow a user
screenshot 2018-12-19 at 23 34 28

### CODE REVIEW

File: src/components/FollowButton/index.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  state = {
+    following: false,
+    loading: false,
+  }
+
+  componentWillMount() {
+    // fetch logged in user following
+    if (isLoggedIn()) {
+      const user = getToken();
+      const { user: { username } } = user;
```
Comment (Rodgers-M) [inline src/components/FollowButton/index.js]: what do you want to achieve in this line?  seems to me that the `user` object has a property `user` that you are trying to rename it to `username`

File: src/components/FollowButton/index.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
+import React, { Component } from 'react';
+import './style.sass';
+import { isLoggedIn, api, getToken } from '../../utils';
+
+export class FollowBtn extends Component {
```
Comment (Rodgers-M) [inline src/components/FollowButton/index.js]: the naming here should be precise, short form is not ideal, you better have a long name but readable.

File: src/components/FollowButton/index.js (hunk 1 above)
Comment (p8ul) [reply src/components/FollowButton/index.js]: @Rodgers-M `getToken()` functions returns an object with this shape 
```
user: {
    username: 'map',
    token: 'token',
  }
```
In line 19, the code is destructuring username from the user object above

Comment (Rodgers-M) [reply src/components/FollowButton/index.js]: you can simply achieve that by 

`const {  username } = user`

File: src/components/FollowButton/index.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+
+  followHandler = () => {
+    const { following } = this.state;
+    const { author } = this.props;
+    this.setState({ loading: true });
+
+    if (!following) {
+      // if am following the author (following == true)
```
Comment (Rodgers-M) [inline src/components/FollowButton/index.js]: you do not need comments, the code should be self-explanatory

File: src/components/FollowButton/index.js (hunk 1 above)
Comment (p8ul) [reply src/components/FollowButton/index.js]: @Rodgers-M Tried it and its not working. You may not destruct username directly from the user object. 
```
var user = {
  user: {
      username: 'map',
      token: 'token',
   }
}

// wrong: outputes "direct distructuring username=>" undefined
let {username} = user;
console.log('direct distructuring username=>', username);

// right
let {user : {username}} = user;
console.log('right distructuring: username=>', username)
```

File: src/components/FollowButton/index.js (hunk 3 above)
Comment (mmosoroohh) [reply src/components/FollowButton/index.js]: okay

File: src/components/FollowButton/index.js (hunk 1 above)
Comment (mmosoroohh) [reply src/components/FollowButton/index.js]: @Rodgers-M We tried that and it got the user object so in order to get the username in the user, you have to destruct the user object where to you get the `username`, `email` and `token`.

**Review by Rodgers-M — APPROVED** [review]

**Pull request merged**